File: app/src/main/java/org/tasks/jobs/RepeatingWorker.java
package org.tasks.jobs;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.WorkerParameters;
import org.tasks.injection.InjectingWorker;

public abstract class RepeatingWorker extends InjectingWorker {

  public RepeatingWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
    super(context, workerParams);
  }

  @NonNull
  @Override
  public final Result doWork() {
    Result result = super.doWork();
    scheduleNext();
    return result;
  }

  protected abstract void scheduleNext();
}

File: app/src/main/java/org/tasks/jobs/RepeatingWorker.kt
package org.tasks.jobs

import android.content.Context
import androidx.work.WorkerParameters
import kotlinx.coroutines.runBlocking
import org.tasks.analytics.Firebase
import org.tasks.injection.BaseWorker

abstract class RepeatingWorker internal constructor(
        context: Context,
        workerParams: WorkerParameters,
        firebase: Firebase
) : BaseWorker(context, workerParams, firebase) {

    override fun doWork(): Result {
        val result = super.doWork()
        runBlocking {
            scheduleNext()
        }
        return result
    }

    protected abstract suspend fun scheduleNext()
}